experimental/regression_suite/ireers_tools/artifacts.py

```diff
@@ -0,0 +1,244 @@
+        """Gets the content_md5 hash for a lolca file, if it exists."""
+        if not local_file_path.exists() or local_file_path.stat().st_size == 0:
+            return None
+
+        with open(local_file_path) as file, mmap.mmap(
+            file.fileno(), 0, access=mmap.ACCESS_READ
+        ) as file:
+            return hashlib.md5(file).digest()
+
+    def download_azure_artifact(self: "FetchedArtifact"):
+        """
+        Checks the hashes between the local file and azure file.
+        """
+        remote_file_name = self.url.rsplit("/", 1)[-1]
+
+        # Extract path components from Azure URL to use with the Azure Storage Blobs
+        # client library for Python (https://pypi.org/project/azure-storage-blob/).
+        #
+        # For example:
+        #   https://sharkpublic.blob.core.windows.net/sharkpublic/path/to/blob.txt
+        #                                            ^           ^
+        #   account_url:    https://sharkpublic.blob.core.windows.net
+        #   container_name: sharkpublic
+        #   blob_name:      path/to/blob.txt
+        result = re.search(r"(https.+\.net)/([^/]+)/(.+)", self.url)
+        account_url = result.groups()[0]
+        container_name = result.groups()[1]
+        blob_name = result.groups()[2]
+
+        with BlobClient(
+            account_url,
+            container_name,
+            blob_name,
+            max_chunk_get_size=1024 * 1024 * 32,  # 32 MiB
+            max_single_get_size=1024 * 1024 * 32,  # 32 MiB
+        ) as blob_client:
+            blob_properties = blob_client.get_blob_properties()
+            blob_size_str = self.human_readable_size(blob_properties.size)
+            azure_md5 = self.get_azure_md5(self.url, blob_properties)
+
+            local_md5 = self.get_local_md5(self.path)
+
+            if azure_md5 and azure_md5 == local_md5:
+                logger.info(
+                    f"  Skipping '{remote_file_name}' download ({blob_size_str}) "
+                    "- local MD5 hash matches"
+                )
+
+            if not local_md5:
+                logger.info(
+                    f"  Downloading '{remote_file_name}' ({blob_size_str}) "
+                    f"to '{self.path}'"
+                )
+                with open(self.path, mode="wb") as local_blob:
+                    download_stream = blob_client.download_blob(max_concurrency=4)
+                    local_blob.write(download_stream.readall())
+            else:
+                logger.info(
+                    f"  Downloading '{remote_file_name}' ({blob_size_str}) "
+                    f"to '{self.path}' (local MD5 does not match)"
+                )
+                with open(self.path, mode="wb") as local_blob:
+                    download_stream = blob_client.download_blob(max_concurrency=4)
+                    local_blob.write(download_stream.readall())
+
+    @staticmethod
+    def _callback(self: "FetchedArtifact"):
+        self.download_azure_artifact()
```

Comment:
Should check that the URL is an Azure URL before calling this function. We'll also want to download from Hugging Face and possibly GCS.